docs/source/llm/run-with-c-plus-plus.md

```diff
@@ -0,0 +1,240 @@
+
+- Manages inputs and outputs to the model
+- Handles KV cache updates
+- Converts logits to tokens via sampling
+
+## Model Metadata
+
+The metadata includes several important configuration parameters:
+
+1. **`enable_dynamic_shape`**: Whether the model supports dynamic input shapes
+2. **`max_seq_len`**: Maximum sequence length the model can handle
+3. **`max_context_len`**: Maximum context length for KV cache
+4. **`use_kv_cache`**: Whether the model uses KV cache for efficient generation
+5. **`use_sdpa_with_kv_cache`**: Whether the model uses the custom op [`torch.ops.llama.sdpa_with_kv_cache.default`](https://github.com/pytorch/executorch/blob/release/0.7/extension/llm/custom_ops/op_sdpa.cpp#L611-L614)
+6. **`get_bos_id`**: Beginning-of-sequence token ID
+7. **`get_eos_ids`**: End-of-sequence token IDs
+8. **`get_vocab_size`**: Size of the model's vocabulary
+
+### Adding Metadata During Export
+
+To ensure your model has the necessary metadata, you can specify it during export using the `metadata` parameter in the export configuration:
+
+```python
+# export_llm
+python -m extension.llm.export.export_llm \
+  --config path/to/config.yaml \
+  +base.metadata='{"get_bos_id":128000, "get_eos_ids":[128009, 128001], "get_max_context_len":4096}'
+```
+
+## Tokenizer Support
+
+The runner library supports multiple tokenizer formats through a unified interface:
+
+```cpp
+std::unique_ptr<tokenizers::Tokenizer> tokenizer = load_tokenizer(
+    tokenizer_path,  // Path to tokenizer file
+    nullptr,         // Optional special tokens
+    std::nullopt,    // Optional regex pattern (for TikToken)
+    0,               // BOS token index
+    0                // EOS token index
+);
+```
+
+Supported tokenizer formats include:
+
+1. **HuggingFace Tokenizers**: JSON format tokenizers
+2. **SentencePiece**: `.model` format tokenizers
+3. **TikToken**: BPE tokenizers
+4. **Llama2c**: BPE tokenizers in the Llama2.c format
+
+For custom tokenizers, you can find implementations in the [pytorch-labs/tokenizers](https://github.com/pytorch-labs/tokenizers) repository.
+
+## Basic Usage Example
+
+Here's a simplified example of using the runner:
+
+```cpp
+#include <executorch/extension/llm/runner/text_llm_runner.h>
+
+using namespace executorch::extension::llm;
+
+int main() {
+  // Load tokenizer and create runner
+  auto tokenizer = load_tokenizer("path/to/tokenizer.json", nullptr, std::nullopt, 0, 0);
+  auto runner = create_text_llm_runner("path/to/model.pte", std::move(tokenizer));
+
+  // Load the model
+  runner->load();
+
+  // Configure generation
+  GenerationConfig config;
+  config.max_new_tokens = 100;
+  config.temperature = 0.8f;
+
+  // Generate text with streaming output
+  runner->generate("Hello, world!", config,
+    [](const std::string& token) { std::cout << token << std::flush; },
+    nullptr);
+
+  return 0;
+}
+```
+
+## Other APIs
+
+1. **Warmup**: For more accurate timing, perform a warmup run before measuring performance:
```

Comment:
Would what this does, assume it's to cold start load the model?

Reply:
For benchmarking purpose